src/build/env.py

```diff
@@ -141,15 +154,23 @@ def __enter__(self):  # type: () -> IsolatedEnvironment
         if 'PATH' in os.environ:
             exe_path.append(os.environ['PATH'])
 
-        self._replace_env('PATH', os.pathsep.join(exe_path))
-        self._replace_env('PYTHONPATH', os.pathsep.join(sys_path))
-        self._replace_env('PYTHONHOME', self.path)
-        self._pop_env('PIP_REQUIRE_VIRTUALENV')
-
         self._symlink_relative(sysconfig.get_path('include'))
         self._symlink_relative(sysconfig.get_path('platinclude'))
         self._symlink_relative(sysconfig.get_config_var('LIBPL'))
 
+        self._symlink_file(sys.executable, os.path.join('bin', 'python'))
+        exe_path.insert(0, os.path.join(self.path, 'bin'))
+
+        self._executable = sys.executable
+        sys.executable = os.path.join(self.path, 'bin', 'python')
+
+        self._replace_env('PATH', os.pathsep.join(sorted(
+            set(exe_path), key=exe_path.index
+        )))  # sorted is required for Python 2 and 3.5
+        self._replace_env('PYTHONPATH', os.pathsep.join(sys_path))
+        self._replace_env('PYTHONHOME', self.path)
+        self._pop_env('PIP_REQUIRE_VIRTUALENV')
```

Comment:
The isolation mechanism is complicated enough that we should have a section that explains how things bound together, and what we're doing to achieve it. (Beside code, architectural level). E.g. Why we need to create bin/python, why we need to alter python path and python home, why we need to pop PIP_REQUIRE_VIRTUALENV. These are all questions we should document within the code. How does pyvenv.cfg is affecting things.

Reply:
You are totally right! That is long overdue. It started simpler but over time things got more complicated. I will document this :blush: